File: src/Common/Message.cs
/* This file implements a common interface for messages that are sent between the client and server.

    Author: Igor Lacko (xlackoi00)
*/

namespace src.Common;

/// <summary>
/// Enum representing the message type
/// </summary>
public enum MessageType
{
    AUTH,
    BYE,
    CONFIRM,
    ERR,
    JOIN,
    MSG,
    PING,
    REPLY,
    NONREPLY, // !REPLY
    EMPTY,
    UNKNOWN
}


/// <summary>
/// Structure representing the message parameters. Contains:
/// <list type="bullet">
///    <item> <description> Username: Mandatory for the AUTH message </description> </item>
///    <item> DisplayName: Mandatory for the AUTH, JOIN, ERR, BYE and MSG messages </item>
///    <item> ChannelID: Mandatory for the JOIN message </item>
///    <item> Secret: Mandatory for the AUTH message </item>
///    <item> MessageContent: Mandatory for the MSG, ERR, REPLY and !REPLY messages </item>
/// </list>
/// </summary>
public struct MessageParams
{
    public string? Username;
    public string? DisplayName;
    public string? ChannelID;
    public string? Secret;
    public string? MessageContent;
}

File: Server/TCP/Message.py
# Contains all messages which can be sent by the client

class AuthMessage:
    def __init__(self, username, displayName, secret):
        self.Username = username
        self.DisplayName = displayName
        self.Secret = secret

    def Parse(message):
        """AUTH ID AS DNAME USING SECRET"""
        split = message.split(" ")
        username = split[1]
        displayName = split[3]
        secret = split[5]
        return AuthMessage(username, displayName, secret)


    def PrintMessage(self):
        print("TYPE: AUTH")
        print(f"USERNAME: {self.Username}")
        print(f"DISPLAYNAME: {self.DisplayName}")
        print(f"SECRET: {self.Secret}")

class JoinMessage:
    def __init__(self, channelID, displayName):
        self.ChannelID = channelID
        self.DisplayName = displayName

    def Parse(message):
        """JOIN ID AS DNAME"""
        split = message.split(" ")
        channelID = split[1]
        displayName = split[3]
        return JoinMessage(channelID, displayName)

    def PrintMessage(self):
        print("TYPE: JOIN")
        print(f"DISPLAYNAME: {self.DisplayName}")

class MsgMessage:
    def __init__(self, displayName, content):
        self.DisplayName = displayName
        self.Content = content

    def Parse(message):
        """MSG FROM DNAME IS CONTENT"""
        split = message.split(" ")
        displayName = split[2]
        content = split[4]
        return MsgMessage(displayName, content)

    def PrintMessage(self):
        print("TYPE: MSG")
        print(f"DISPLAYNAME: {self.DisplayName}")
        print(f"CONTENT: {self.Content}")

class ErrMessage:
    def __init__(self, displayName, content):
        self.DisplayName = displayName
        self.Content = content

    def Parse(message):
        """ERR FROM DNAME IS CONTENT"""
        split = message.split(" ")
        displayName = split[2]
        content = split[4]
        return ErrMessage(displayName, content)

    def PrintMessage(self):
        print("TYPE: ERR")
        print(f"DISPLAYNAME: {self.DisplayName}")
        print(f"CONTENT: {self.Content}")

class ByeMessage:
    def __init__(self, displayName):
        self.DisplayName = displayName

    def Parse(message):
        """BYE FROM DNAME"""
        split = message.split(" ")
        displayName = split[2]
        return ByeMessage(displayName)

    def PrintMessage(self):
        print("TYPE: BYE")
        print(f"DISPLAYNAME: {self.DisplayName}")